src/wasm/wasm-type.cpp

```diff
@@ -2203,6 +2291,258 @@ bool FiniteShapeEquator::eq(const Rtt& a, const Rtt& b) {
   return a.depth == b.depth && eq(a.heapType, b.heapType);
 }
 
+bool RecGroupStructure::operator==(const RecGroupStructure& other) const {
+  if (group == other.group) {
+    return true;
+  }
+  // The rec groups are equivalent if they are piecewise equivalent.
+  return std::equal(group.begin(),
+                    group.end(),
+                    other.group.begin(),
+                    other.group.end(),
+                    [&](const HeapType& a, const HeapType& b) {
+                      return topLevelEq(a, b, other.group);
+                    });
+}
+
+bool RecGroupStructure::topLevelEq(HeapType a,
+                                   HeapType b,
+                                   RecGroup other) const {
+  if (a == b) {
+    return true;
+  }
+  if (a.isBasic() || b.isBasic()) {
+    return false;
+  }
+  return eq(*getHeapTypeInfo(a), *getHeapTypeInfo(b), other);
+}
+
+bool RecGroupStructure::eq(Type a, Type b, RecGroup other) const {
+  if (a == b) {
+    return true;
+  }
+  if (a.isBasic() || b.isBasic()) {
+    return false;
+  }
+  return eq(*getTypeInfo(a), *getTypeInfo(b), other);
+}
+
+bool RecGroupStructure::eq(HeapType a, HeapType b, RecGroup other) const {
+  // Do not recurse into the structure of children `a` and `b`, but check
+  // whether their recursion groups and indices match. Since the wrapped group
+  // may not be canonicalized, explicitly check whether `a` and `b` are in the
+  // respective recursion groups of the respective top-level groups we are
+  // comparing, in which case the structure is still equivalent.
+  if (a.getRecGroupIndex() != b.getRecGroupIndex()) {
+    return false;
+  }
+  auto groupA = a.getRecGroup();
+  auto groupB = b.getRecGroup();
```

Comment:
If `other` is the group of B, why do we need this line? The next line even compares the two. So is `other` not simply the group of `b`, piped down into all these `eq()` functions - I thought that's what you said before?

Reply:
`other` is the rec group that `b` is being used in, not necessarily the rec group that `b` belongs to.

Reply:
I think this might be the core of my misunderstanding then. So this method is called where a is used in the wrapped group, and b is used in the `other` group, but they both might belong to other groups? Or is that true only for b and not a?

Maybe adding concrete examples of rec groups in a comment might help. And/or, a comment that explains when these eq methods are used - what are the presuppositions about groups and canonicalization state and so forth are (e.g., "b must belong to .., but a is still being"). This may be super-obvious to you, sorry about that! But the isorecursive semantics are odd enough for me that I can't just get it from reading this code I guess.

Reply:
Also, perhaps `other => bUser` to clarify it is the group that uses `b`?